src/test/controller/UserController.test.ts

```diff
@@ -399,4 +375,59 @@ describe('Unit tests for UserController', function () {
 
     expect(actualUser).toMatchObject(expectedObject);
   }
+
+  /**
+   * Save the promotion and user to the database. Also add the saved promotion into the database
+   * * Set transaction level to serializable and set lock level to pessimistic write to avoid deadlocks
+   * */
+  async function addSavedPromotion(user: User, promotion?: Promotion) {
+    await getManager().transaction(
+      'SERIALIZABLE',
+      async (entityManager: EntityManager) => {
+        return entityManager
+          .createQueryBuilder()
+          .setLock('pessimistic_write')
+          .insert()
+          .into(User)
+          .values(user)
+          .execute();
+      }
+    );
+
+    if (promotion) {
+      await getManager().transaction(
+        'SERIALIZABLE',
+        async (entityManager: EntityManager) => {
+          return entityManager
+            .createQueryBuilder()
+            .setLock('pessimistic_write')
+            .insert()
+            .into(Promotion)
+            .values(promotion)
+            .execute();
+        }
+      );
+
+      await getManager().transaction(
+        'SERIALIZABLE',
+        async (entityManager: EntityManager) => {
+          return entityManager
+            .createQueryBuilder()
+            .setLock('pessimistic_write')
+            .insert()
+            .into(SavedPromotion)
+            .values([
+              {
+                promotionId: promotion.id,
+                userId: user.id,
+              },
+            ])
+            .execute();
+        }
+      );
+    }
+
+    // This is an unfortunate hack...need to invest more time into this
+    await new Promise((resolve) => setTimeout(resolve, 500));
```

Comment:
Sorry, maybe I'm missing this but where do you do step (4) here? i.e. GET or DELETE saved promotion

Is it the `connection.clear()`?

Reply:
Oh step (4) is in the actual test case itself when we use `supertest` to test our endpoints